io-sim/src/Control/Monad/IOSim/Internal.hs

```diff
@@ -695,29 +740,37 @@ reschedule !simstate@SimState{ threads, timers, curTime = time } =
         -- Reuse the STM functionality here to write all the timer TVars.
         -- Simplify to a special case that only reads and writes TVars.
         !written <- execAtomically' (runSTM $ mapM_ timeoutAction fired)
-        (wakeup, wokeby) <- threadsUnblockedByWrites written
+        (wakeupSTM, wokeby) <- threadsUnblockedByWrites written
         !_ <- mapM_ (\(SomeTVar tvar) -> unblockAllThreadsFromTVar tvar) written
 
-        let (unblocked,
-             simstate') = unblockThreads wakeup simstate
+            -- Check all fired threadDelays
+        let !wakeupThreadDelay = mapMaybe threadDelayThreadId fired
```

Comment:
For this I''d just use a list comprehension, rather than a separate named function. e.g.
```
let wakeupThreadDelay = [ tid | TimerThreadDelay tid <- fired ]
```
Note also that because it's a list, there's no point banging it.

Reply:
Done